docs/architecture/adr-110-remote-mempool.md

```diff
@@ -0,0 +1,364 @@
+   prevent DDoS attacks.
+2. The application validates incoming transactions internally in a similar way
+   to how `CheckTx` currently works.
+3. Transactions that the application deems valid are then sent to the mempool by
+   the application, e.g. through some form of RPC mechanism. The mempool is then
+   expected to propagate these transactions to the rest of the network.
+4. During `PrepareProposal`, a Comet validator will reap transactions from the
+   mempool by way of an RPC call.
+
+Additionally, when transactions are received by the local mempool instance from
+foreign mempool instances, it is expected that the local mempool instance will
+send the transaction to the application for validation (similar to how `CheckTx`
+currently functions).
+
+#### Scalable mempool
+
+The recommended high-level architecture for a modular Comet validator with an
+externalized, scalable mempool, is as follows.
+
+```mermaid
+flowchart LR
+    comet[Comet Validator]
+    app[Application]
+    mempoolrouter[Mempool Router]
+    mempool1[Mempool Instance 1]
+    mempool2[Mempool Instance 2]
+    mempoolN[Mempool Instance N]
+    user[User]
+
+    comet --> app
+
+    comet -- 4. Reap transactions --> mempoolrouter
+    user -- 1. Submit transactions --> app
+    app -- 2. CheckTx --> app
+    app -- 3. Publish transactions --> mempoolrouter
+
+    mempoolrouter --> mempool1
+    mempoolrouter --> mempool2
+    mempoolrouter --> mempoolN
+
+    mempool1 -- CheckTx --> app
+    mempool2 -- CheckTx --> app
+    mempoolN -- CheckTx --> app
+```
+
+Here the interaction is the same as the simple variant discussed earlier, but
+instead of interacting directly with the mempool, all interactions are with some
+form of routing mechanism. For example, this could be a reverse proxy like nginx
+configured to deliver requests to mempool instances in a round-robin fashion.
+
+This configuration would be application-specific, and would need to be set up
+correctly by the operator for the specific application.
+
+### Configuration
+
+The following change to the `config.toml` file is envisaged:
+
+```toml
+[mempool]
+# The type of mempool for this CometBFT node to use.
+#
+# Valid types of mempools supported by CometBFT:
+# - ""       : Default clist mempool with flooding gossip protocol
+# - "remote" : Remote mempool in a separate process
+type = "remote"
+
+#
+# Configuration specific to the remote mempool. If mempool.type is not "remote",
+# this section will be ignored.
+#
+# A remote mempool is only usable by a validator node. Turning on the remote
+# mempool for a full node will simply disable any mempool-related functionality
+# on that full node, and the full node will not interact with any mempool at
+# all.
+#
+[mempool.remote]
+# The base URL for the gRPC interface to the remote mempool.
+url = "http://localhost:28880"
+
+# The timeout for fetching transaction data during PrepareProposal. Exceeding
+# this timeout will result in a panic.
```

Comment:
I agree that the proposer should continue operating, even with a misconfigured mempool, but trying again most probably won't fix the problem and it delays block creation. I think it should let other validators to propose.

Reply:
Sure, we could do that. I thought about this afterwards that we could actually just use the consensus proposal timeout as the upper bound here instead of defining a dedicated parameter.